sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,93 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef ITYPE_t next_label
+    cdef ITYPE_t[:] parent
+    cdef ITYPE_t[:] size
+
+    def __init__(self, N):
+        self.parent = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size = np.hstack((np.ones(N, dtype=ITYPE),
+                               np.zeros(N - 1, dtype=ITYPE)))
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.parent[m] = self.next_label
+        self.parent[n] = self.next_label
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.next_label += 1
+
+        return
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef ITYPE_t fast_find(self, ITYPE_t n):
+        cdef ITYPE_t p
+        p = n
+        # find the highest node in the linkage graph so far
+        while self.parent[n] != -1:
+            n = self.parent[n]
+        # provide a shortcut up to the highest node
+        while self.parent[p] != n:
+            p, self.parent[p] = self.parent[p], n
+        return n
+
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] single_linkage_label(
+    np.ndarray[DTYPE_t, ndim=2] L):
+    """
+    Convert an linkage array or MST to a tree by labelling clusters at merges.
+    This is done by using a Union find structure to keep track of merges
+    efficiently.
+
+    Parameters
+    ----------
+    L: array of shape (n_samples - 1, 3)
+        The linkage array or MST where each row specifies two samples
+        to be merged and a distance or weight at which the merge occurs. This
+         array is assumed to be sorted by the distance/weight.
+
+    Returns
+    -------
+    A tree in the format used by scipy.cluster.hierarchy.
+    """
+
+    cdef np.ndarray[DTYPE_t, ndim=2] result_arr
+    cdef DTYPE_t[:, ::1] result
+
+    cdef ITYPE_t left, left_cluster, right, right_cluster, index
+    cdef DTYPE_t delta
+
+    result_arr = np.zeros((L.shape[0], 4), dtype=DTYPE)
+    result = (<DTYPE_t[:L.shape[0], :4:1]> (
+        <DTYPE_t *> result_arr.data))
```

Comment:
Is explicit casting needed here? I think assigning the array itself to an appropriately cdefed memoryview is sufficient.

Reply:
That makes sense yes, I was borrowing some old code I wrote before I was suitably familiar with memory views.